### PR TITLE
[file_selector] Add getDirectoryPaths implementation on macOS

### PR DESCRIPTION
**Pull request by VanesaOshiro**
This PR adds the macOS implementation for retrieving multiple directories paths from a select folder dialog.

Issue: [Support for selection of multiple directories, through desktop's native open panel, in 'file_selector' package](https://github.com/flutter/flutter/issues/74323)

image
_New option on example application_

image
_GetDirectoriesPaths_

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by VanesaOshiro** [comment]
Changes applied, thanks!

File: packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift:76]: You've correctly changed the Dart side to expect a list, but are not sending one, so this will crash.

File: packages/file_selector/file_selector_platform_interface/lib/src/method_channel/method_channel_file_selector.dart
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/lib/src/method_channel/method_channel_file_selector.dart:98]: There's no such method, so this won't work.

**Review by stuartmorgan — COMMENTED** [review]
The comment sounded like this was ready for review, but it seems like it's still actually a draft since it doesn't appear this version was ever run for manual testing. Was this waiting for something on the review side before being completed?

File: packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift
Comment (adpinola) [reply packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift]: Change applied, thanks.

**Comment by adpinola** [comment]
Hi @stuartmorgan, referring to your last comment:

>The comment sounded like this was ready for review, but it seems like it's still actually a draft since it doesn't appear this version was ever run for manual testing. Was this waiting for something on the review side before being completed?

This PR is still in draft because it depends on #6572, the same occurs with #6573 which depends also on #6572.
#6572 has your approval but we are still waiting on the second one, so we can merge that and set this ready for review.

Thanks!

File: packages/file_selector/file_selector_macos/CHANGELOG.md
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/CHANGELOG.md:1]: This should be 0.9.1 since it's a new feature.

File: packages/file_selector/file_selector_macos/example/lib/get_multiple_directories_page.dart
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/example/lib/get_multiple_directories_page.dart:27]: Per comment in another implementation's review, it seems like this should just be `directoryPaths.join('\n')`.

File: packages/file_selector/file_selector_macos/example/lib/get_multiple_directories_page.dart
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/example/lib/get_multiple_directories_page.dart:20]: This can't be null.

File: packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/macos/Classes/FileSelectorPlugin.swift:76]: This needs to be re-indented.

File: packages/file_selector/file_selector_macos/lib/file_selector_macos.dart
Comment (stuartmorgan) [inline packages/file_selector/file_selector_macos/lib/file_selector_macos.dart:89]: This and `openFile` are actually both fragile; they are assuming cancel will be null rather than an empty array, which isn't something we should count on.

Let's add a helper function to use in both:
```
String? firstOrNull(List<String>? list) {
  return (list?.isEmpty ?? true) ? null : list.first;
}
```

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by stuartmorgan** [comment]
What's the state of this PR? Is it ready for re-review?

**Comment by adpinola** [comment]
>What's the state of this PR? Is it ready for re-review?

Yes, it is.

**Comment by stuartmorgan** [comment]
Apologies for the churn, but this will need to be reworked for the just-landed https://github.com/flutter/plugins/pull/6902. I forgot that this was in flight when doing that migration, or I would have waited for this to land first.

**Comment by stuartmorgan** [comment]
@VanesaOshiro Are you still planing on updating this PR?

**Comment by VanesaOshiro** [comment]
Hi @stuartmorgan. We closed this PR and [opened a new one](https://github.com/flutter/plugins/pull/7115) integrating the use of Pigeon.
Please feel free to review it.

Thanks!

**Pull request closed without merging**
